jupyter_archive/handlers.py

```diff
@@ -223,17 +223,42 @@ async def get(self, archive_path, include_body=False):
         self.finish()
 
     def extract_archive(self, archive_path):
+        def try_macos_decode(fn):
+            try:
+                return fn.encode('cp437').decode('utf-8')
+            except UnicodeError:
+                return None
+
+        def try_windows_chinese_decode(fn):
+            try:
+                return fn.encode('cp437').decode('gbk')
+            except UnicodeError:
+                return None
+
+        def unzip(reader, destination):
+            # most of Windows implementations use DOS (OEM) encoding
+            # Mac OS zip utility uses utf-8, but it doesn't set utf-8 bit flags
+            # *nix zip utilities silently uses system encoding(utf-8 generally)
+            with reader as f:
+                for fn in f.namelist():
+                    extreact_path = pathlib.Path(f.extract(fn, path=destination))
+                    correct_filename = try_macos_decode(fn) or try_windows_chinese_decode(fn) or fn
```

Comment:
Could you change this a bit to use a single function that has the following structure:

```py
def decode_filename(fn):
    try:
        encoded_filename = fn.encode('cp437')
        try:
            # MacOS encoding
            return encoded_filename.decode('utf-8')
        except UnicodeError:
            # Windows encoding
            return encoded_filename.decode('gbk')
    except UnicodeError:
        return fn
```